### PR TITLE
doc: add qemu tutorial

### PR DESCRIPTION
**Pull request by holmanb**
```
doc: add qemu tutorial
```

## Additional Context
The [NoCloud docs](https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html) for more information about the datasource in use in this tutorial.

When more tutorials come into existance, a landing page/tutorial index for selecting which datasource to use may become desirable, but at only 2 total tutorials as of this PR, I don't know how much sense that would make. Thoughts?

### CODE REVIEW

**Comment by cpaelzer** [comment]
Hey,
generally LGTM, but I've left a few nit picks to improve this or start a discussion.

**Comment by TheRealFalcon** [comment]
@cpaelzer  ` but I've left a few nit picks`. They're not here. Did you comment from this page rather than the "Files changed" tab?

File: doc/rtd/index.rst
Comment (s-makin) [inline doc/rtd/index.rst:47]: I'm a bit confused by the doc structure here. Is the Qemu tutorial subordinate to the LXD one, or are they just different approaches to achieving the same end? If the latter,they should really be at the same level in the hierarchy. In that case, I'd say yes, absolutely we need a tutorials landing page (although I'm intending to make one as part of our Diataxification efforts).

Although then I would have to ask "do we need two tutorials on the same topic?". Is one of these methods more common, easier to use, or otherwise preferable in some way? If so, that should be what we use to get people started, and the other tutorial may be better as a how-to guide.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:3]: We may need to make the title a bit more descriptive. Where it appears in the LHS menu as just "Qemu", a new user may not know what it is or whether it's relevant to their interests.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:6]: I'd rework this paragraph a bit to first briefly explain what Qemu is and why someone might want to use it for deploying cloud-init, and then telling the user what end goal the tutorial will walk them towards.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:16]: Is it conventional to use the $ signs in the code snippets? A couple of times they caused me problems when I was copy/pasting them.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:18]:
```suggestion
If you are not using Ubuntu, you can visit Qemu's `install instructions <https://www.qemu.org/download/#linux>`_ for additional information.
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:23]: If it's generally considered a good idea to create a temporary directory, then for a tutorial doc this should be presented as an explicit step rather than an optional one. It wasn't completely clear to me if I had to download the image into the temp directory, so I didn't create one. It didn't occur to me until after that maybe I should have done that.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:37]:
```suggestion
Create a file called ``user-data`` in your temporary directory and copy/paste the following code into it.
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:66]:
```suggestion
Now create the empty file ``vendor-data`` in your temporary directory. This will speed up the retry wait time.
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:73]: I have no idea what an IMDS Server is or why I'd want one - this section needs a bit of context if it's aimed at a new user.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:81]:
```suggestion
Launch a VM with our user data
```

Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:81]:
```suggestion
Launch a virtual machine (VM) with our user data
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:103]: It wasn't clear to me that I had to press enter after the VM was launched to make the login bit appear

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:121]:
```suggestion
Exit the qemu shell using ``ctrl-a x`` (that's ``ctrl`` and ``a``
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:125]: I think we're supposed to avoid using words like "kill" - is there an alternative we can use here (cancel, quit, exit, etc).

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:132]: Did we do that with a module?

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:135]:
```suggestion
The documentation for each module contains examples of how to use it.
```

**Review by s-makin — CHANGES_REQUESTED** [review]
Sadly I wasn't able to complete the tutorial fully - it didn't like the password, even when I tried to copy/paste it. 

I have some comments, questions and suggestions as well, which I've left on the docs.

File: doc/rtd/topics/tutorials/lxd.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/lxd.rst:10]: Correct

File: doc/rtd/index.rst
Comment (holmanb) [reply doc/rtd/index.rst:47]: Since developers' preferred virtualization "test environment" tooling varies from person to person (qemu, libvirt, lxd, virtualbox, vagrant, etc). I think it is desirable to provide multiple tutorials.

Additionally, Qemu is capable of running virtual machines on [BSD](https://wiki.qemu.org/Hosts/BSD) / [MacOS](https://wiki.qemu.org/Hosts/Mac), while [LXD does not](https://ubuntu.com/blog/lxd-client-on-windows-and-macos), which means that this tutorial may allow some users to test locally despite lacking LXD support.

I think a landing page makes sense, but with only 2 tutorials I wasn't sure the best way to go about that (a list with only 2 items seems a bit hollow).

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:20]: I don't believe so.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:92]: > do you strictly need q35 type?

No

> almost everywhere "as fast as it can"

As a tutorial I think compatibility with various realease/distros is more important than speed. Fallback to tcg makes sense here. +1 thanks

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:37]: I've made this confusing, sorry.

The text here describes what the code does, not what the reader is supposed to do. If one creates a file named `user-data` and then copies in the following contents, the tutorial will not work.

The tutorial is written such that all the user should have to do is copy, paste, and execute each code block.

I'll add some clarifying instructions at the top to make this clear.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:52]: Same comment as [this](https://github.com/canonical/cloud-init/pull/1863/files#r1034030434).

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:66]: +1 I like the clarity this adds.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:84]: Good catch! I'll fix that. LXD shouldn't be mentioned in this tutorial.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:103]: Yeah that's confusing? I forgot about that.

I think what's happening there is that when the prompt is ready it prints to stdout, but since systemd and kernel logs are still printing the old prompt gets lost in the mix (if you scroll up you might see it). Stdin is listening, even if you don't see the prompt, but hitting enter will make it print again.

I'll add something to make it clear to users.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:107]: I think that [this comment](https://github.com/canonical/cloud-init/pull/1863/files#r1034030434) might explain what happened, however you raise a good point that some help for how to debug this would be helpful.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:125]: +1

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:132]: I'll add a comment.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:3]: +1

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:6]: +1

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:16]: Yes, this is a convention. It helps to disambiguate whether a command is single-line or multi-line, and is sometimes used to indicate non-root vs root commands (using $ and #, respectively).

Consider the heredoc commands below (line 41). The $ here helps users to understand that these are not individual commands to be executed in sequence, but are part of a multi-line shell command.

I've seen this question come up enough times in IRC/Stack Exchange/etc that I agree it is potentially worth dropping to avoid confusion. If not for the heredoc commands (see line 41), I think this would be easy to eliminate them all.

My concern is that if I saw the following while skimming a *nix tutorial, I would have a moment of pause because lines after the first one are not valid shell commands. 
```
    cat << EOF > user-data
    #cloud-config
    password: passw0rd
    chpasswd:
      expire: False
    EOF
```
Is this more readable to others? If we're alright with this format I'd be fine with dropping all of the $s.

Thoughts?

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:23]: +1

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:57]: +1

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:73]: Fair enough :D

File: doc/rtd/index.rst
Comment (s-makin) [reply doc/rtd/index.rst:47]: It's an interesting problem. Diataxis' end goal is a single end-to-end tutorial, although that may or may not be appropriate here. 

The key is going to be signposting, I think. If the methods really are equivalent but the difference is *who* might use each one, we need to include some context around that decision so that people can pick the right tutorial for their needs without having to skim each one first (saving time and effort for the user). E.g., someone who wants to take part in development (I'm guessing) has a different set of needs than a sysadmin. If each tutorial is better for a different purpose, we can (and probably should) state that explicitly to make it clear, or at least provide the brief explanation you just made as an introduction on the tutorial landing page (which we will need) to explain the two options.

Out of curiosity, is there a tool that would be ideal for non-dev/non-sysadmin types who want to try cloud-init?

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst:37]: This is probably how I managed to break the tutorial, to be fair...lol

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst:107]: It may not be necessary to add debug info, if you're already making the step I messed up clearer (if we can avoid someone doing the stupid thing I did, they wouldn't need to debug in the tutorial)

File: doc/rtd/topics/tutorials/qemu.rst
Comment (aciba90) [reply doc/rtd/topics/tutorials/qemu.rst:38]: It does also indicate what lines are command and what are output:

```
$ pwd
/home/aciba
```

And prevents from disaster while copying and pasting commands, if I copy-paste `$ rm /` and press enter, then I will get an error because of the invalid `$`.

File: doc/rtd/index.rst
Comment (holmanb) [reply doc/rtd/index.rst:47]: > is there a tool that would be ideal for non-dev/non-sysadmin types who want to try cloud-init?

Hmmm, Virtualbox probably most "cross-platform", but Qemu is more widely used on Linux. I don't have a good answer to this.

**Comment by holmanb** [comment]
Thanks @s-makin and @cpaelzer for the reviews.

I believe that I have addressed all requested changes. Requesting re-review.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [inline doc/rtd/topics/tutorials/qemu.rst:87]: Oops, forgot this note to self (will add a minor patch for this tomorrow).

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu-debugging.rst:1]:
```suggestion
.. _qemu_debug_info:
```

Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu-debugging.rst:1]: For consistency with all the other links I'd stick an underline in there. Also, debug information seems like it could have broad applications, so I'd also recommend making the tag more specific to the page (in case we need to use "debug information" for a more general page in the future)

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu-debugging.rst:3]:
```suggestion
Qemu tutorial debugging
```

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu-debugging.rst:14]:
```suggestion
Did cloud-init discover the IMDS webserver?
```

Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu-debugging.rst:14]: (You capitalised the acronym previously, but I'm not sure which is correct - please disregard this comment/suggestion if it shouldn't be capitalised)

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu-debugging.rst:29]:
```suggestion
Did the IMDS webserver serve the expected files?
```

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu-debugging.rst:38]:
```suggestion
Were the configurations inside the file correct?
```

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu-debugging.rst:35]:
```suggestion
If the webserver prints out 404 errors when launching Qemu, then first check that you started the server in the temp directory. 

```

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu-debugging.rst:38]: Also, just want to double check what file "the file" refers to here?

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst:21]: Nice - this is exactly what I was looking for :D just be careful not to switch from talking to the user (as in the first sentence) to talking about them (in this one).

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:27]: This is really good - I would just put a link to the Qemu website here at this first proper mention of it (for convenience)

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:41]: Not sure if this is my lack of knowledge here - I'm a bit confused by the use of "surface" as a verb. Is this a common terminology that we can assume the users will know/understand?

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:50]:
```suggestion
The IMDS contains a private http webserver to each operating
```

Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:50]: "contains to" feels a bit odd to my eyes. Again I'm not sure if this is my lack of knowledge at play.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:157]:
```suggestion
what kind of configuration is contained. The cloud-config config type use yaml
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:211]:
```suggestion
Launch the virtual machine. By default, Qemu will print the kernel logs
and systemd logs to the terminal while the operating system boots. This
may take a few moments to complete.
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:213]:
```suggestion
If the output stopped scrolling but you don't see a prompt yet, press ``enter``
```

Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst:213]: I got a bit confused by the code block appearing below this. I'd swap the order (so the thing about the prompt appears below the code block).

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:242]:
```suggestion
:ref:`this page for debug information<qemu_debug_info>`.
```

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [inline doc/rtd/topics/tutorials/qemu.rst:14]:
```suggestion
```

Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst:14]: This looks a bit odd when rendered in RTD because it's so close to the page title. I'd recommend just taking it out, since it's not really needed.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst:211]: Just to make it completely crystal clear, I'd preface this little bit by saying something like "In our original terminal window" or something like that. Idiot that I am, I sat wondering which I was meant to use for a good minute until I realised that I couldn't use the IMDS window while the server was running.

**Review by s-makin — COMMENTED** [review]
It's looking really good! The parts you added to give context are awesome; it feels much clearer now when I'm reading it and I understand a lot more about what's happening. I especially like how you echoed the aims of the tutorial from the start with the "what we achieved" summary at the end, like tying up a loose end. I have a few little questions and suggestions, but overall I think it's great :D

As an added bonus, I managed to complete it successfully without breaking it this time!

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu-debugging.rst:38]: Good point. This refers to the user-data configuration file. I'll add that.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:39]: **I** use this term a lot, though this is probably more colloquial than it needs to be. I'll use "to expose" instead.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:48]: +1

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:155]: I'm going to push back a little on this change request. Since this config "type" is singular wouldn't `use` make more sense here?

I'm happy to make changes for clarity, but I do think that the current grammar makes more sense (which leads me to think that something about this sentence was confusing).

Thoughts?

Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst:155]: You're quite right, grammatically your original wording is correct, I have no idea why I thought it wasn't - I think the double "config" might have made my brain short circuit for a hot minute. If there is an alternative way of phrasing that sentence without changing your meaning, it might be a good idea to change it. It's a bit confusing to look at.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst:25]: +1

File: doc/rtd/topics/tutorials/qemu.rst
Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu.rst]: > If there is an alternative way of phrasing that sentence without changing your meaning, it might be a good idea to change it. It's a bit confusing to look at.

How is this wording? https://github.com/canonical/cloud-init/pull/1863/commits/8398443f79ca46ce25beee74cb7f0eba23046f49

**Comment by holmanb** [comment]
> It's looking really good! The parts you added to give context are awesome; it feels much clearer now when I'm reading it and I understand a lot more about what's happening. I especially like how you echoed the aims of the tutorial from the start with the "what we achieved" summary at the end, like tying up a loose end. I have a few little questions and suggestions, but overall I think it's great :D

Thanks for the reviews @s-makin, they've been really helpful to make sure we're not glossing over details or assumed knowledge.

> As an added bonus, I managed to complete it successfully without breaking it this time!

:-)

**Comment by holmanb** [comment]
I added one more section titled `How is Qemu configured for cloud-init?`, to direct the reader to the important cloud-init-specific parts of the command. Since it uses a specific IP address and port that isn't mentioned anywhere in the tutorial, it desires some explanation.

I think this is ready for re-review @s-makin.

File: doc/rtd/topics/tutorials/qemu.rst
Comment (s-makin) [reply doc/rtd/topics/tutorials/qemu.rst]: Much better! Clear, but still concise - good job +1

File: doc/rtd/index.rst
Comment (s-makin) [reply doc/rtd/index.rst:47]: After this PR is merged, I'll be creating Diataxis section landing pages, so I can easily link both tutorials from the Tutorial landing page (fixing the navigation issue). If it means we have two tutorials for a while, I'm totally OK with that. My personal preference would be to use the LXD one as a quick-start tutorial since it could easily fill that gap as-is. I completed it before np, so I think it works well for the subset of users who don't need introducing to the concepts. 

For the time being, I could simply include introductory text on the Tutorial landing page to direct users to the tutorial that they want to use? (quick start LXD vs detailed Qemu). It's a tidy option, and might end up working really well from the users' point of view. In this way we have two bases covered, and serve two different subsets of users with the Tutorial content.

Comment (TheRealFalcon) [reply doc/rtd/index.rst:47]: +1, sounds good to me!

File: doc/rtd/topics/tutorials/qemu-debugging.rst
Comment (TheRealFalcon) [inline doc/rtd/topics/tutorials/qemu-debugging.rst:45]: Should this be an internal link?

Comment (holmanb) [reply doc/rtd/topics/tutorials/qemu-debugging.rst]: Thanks! Fixed in c77b0579c8363929de6

File: doc/rtd/index.rst
Comment (holmanb) [reply doc/rtd/index.rst:47]: Thanks all for the reviews! I don't see any unaddressed issues[1]. I'll merge once tests pass, barring any last minute objections.

[1] Thanks @s-makin for offering to pick up the navigation with Diataxification!!

**Comment by holmanb** [comment]
> Nicely done. I think this does a great job conveying some of the concepts that are central to cloud-init.

Thanks :-)

> I do think this tutorial requires more complicated manual steps like downloading an image, setting up an IMDS, and all of the various QEMU args, along with a slower boot time. I understand that this helps explain some concepts behind cloud-init better, but if this becomes the default tutorial, I would like to see a pathway in our docs for somebody that wants the simplest way to get a "try, check, try again" loop.

+1 Agreed.

**Pull request merged**
